ssl/quic/quic_rx_depack.c

```diff
@@ -0,0 +1,612 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <internal/packet.h>
+#include <internal/nelem.h>
+#include <internal/quic_wire.h>
+#include <internal/quic_record_rx.h>
+#include <internal/quic_ackm.h>
+
+#include "quic_local.h"
+#include "../ssl_local.h"
+
+/*
+ * ASSUMPTION: the QUIC_CONNECTION structure refers to other related
+ * components, such as OSSL_ACKM and OSSL_QRX, in some manner.  These macros
+ * should be used to get those components.
+ */
+#if 0                            /* Currently unimplemented */
+# define GET_CONN_ACKM(c) ((c)->ackm)
+#else
+# define GET_CONN_ACKM(c) NULL
+#endif
+#if 0                            /* Currently unimplemented */
+# define GET_CONN_QRX(c) ((c)->qrx)
+#else
+# define GET_CONN_QRX(c) NULL
+#endif
+#if 0                            /* Currently unimplemented */
+# define GET_CONN_STATM(c) ((c)->statm)
+#else
+# define GET_CONN_STATM(c) NULL
+#endif
+
+/* =====BEGIN TO BE REMOVED===== */
+
+/* Diverse things that should be implemented elsewhere, bur currently aren't. */
+
+typedef struct quic_stream_st QUIC_STREAM;
+
+/*
+ * ASSUMPTION: ssl_get_stream() gets a QUIC_STREAM from a connection by
+ * stream ID.
+ */
+static QUIC_STREAM *ssl_get_stream(QUIC_CONNECTION *conn, uint64_t stream_id)
+{
+    return NULL;
+}
+
+/* ASSUMPTION: ssl_get_stream_tyoe() gets a stream type from a QUIC_STREAM */
+#define SSL_STREAM_TYPE_R       1
+#define SSL_STREAM_TYPE_W       2
+#define SSL_STREAM_TYPE_RW      (SSL_STREAM_TYPE_R|SSL_STREAM_TYPE_W)
+static int ssl_get_stream_type(QUIC_STREAM *stream)
+{
+    return SSL_STREAM_TYPE_RW;
+}
+
+/*
+ * ASSUMPTION: ssl_queue_data() adds data to an QUIC_STREAM, to be consumed
+ * by the application when doing an SSL_read().
+ *
+ * It's currently not determined what that queueing does, if it copies the
+ * data into another buffer, or if that should attach the OSSL_QRL_RX_PKT to
+ * the stream SSL queue, and exactly what that entails in terms of deallocation
+ * of that packet.
+ */
+static int ssl_queue_data(QUIC_STREAM *stream,
+                          const unsigned char *data, uint64_t data_len)
```

Comment:
This looks too simplistic to me. I don't think we should be copying data here - just copying pointers to data. That means the underlying buffer needs to remain available for the whole time that the data sits in the stream queue. I suggest we pass the `OSSL_QRX_PKT` object and this function takes a reference to it. Only when the reference drops to 0 would the packet be released and the underlying buffer removed.

Reply:
Sure.  Do note that `ssl_queue_data()` is simply a placeholder for whatever the function is actually going to be called, and what parameters it's going to take.  You want a `OSSL_QRX_PKT` parameter?  You got it!

Reply:
There, parameter added, and comment enhanced.